proposals/0001-metadata-providers.md

```diff
@@ -0,0 +1,328 @@
+access in order to obtain the `Baggage` value, rather than the lookup being able to be performed _once_ and shared between 
+multiple providers when a multiplex provider was configured in the system. We view this tradeoff as worth taking, as the cost
+of actually formatting the metadata usually strongly dominates the cost of the task-local lookup. 
+
+### Removing `LogHandler.Metadata` in favor of `Baggage`
+
+Removing logger metadata is not a good option because it serves a slightly different style of metadata than the baggage.
+
+Baggage is intended for contextual, task-local, metadata which is _carried across multiple libraries_. The values stored in baggage are well-typed, and must declare keys for accessing values in a baggage, this works well for multiple pieces of an application needing to reach for specific baggage items: everyone aware of the `traceID` key, is able to query the baggage for this key (`baggage.traceID`) and obtain a well-typed value. This comes with a higher cost on declaring keys though, as well as a global namespace for those - which is desirable for such kinds of metadata.
+
+This is not the same usage pattern as emitting a plain structured log where we'd like to include the name of an item we just queried:
+
+```swift
+log.info("Obtained item! Hooray!", metadata: ["item": "\(item)"])
+```
+
+In this example, the key/value pair for `"item"` is pretty ad-hoc, and we never need to refer to it elsewhere in the program. It never is queried by other pieces of the application, nor would it be useful to set it in baggage metadata, as the only purpose of the `item` key here is to log, and forget about it.
+
+### Explicitly passing `Baggage` always
+
+Baggage is designed for use cases like distributed tracing, or similar instrumentation patterns where "all" participating code may need to reach for it.
+
+Specifically in logging, this means that _every_ call site for _every_ log statement would have to pass it explicitly resulting in annoying noisy code:
+
+```swift
+class StoresRepository {
+    func store(byID id: String, eventLoop: EventLoop, logger: Logger, baggage: Baggage) async throws -> Store {
+        InstrumentationSystem.tracer.withSpan("Fetch Store") { span in
+            logger.info("Fetched store.", baggage: span.baggage)
+        }
+    }
+}
+
+try await storesRepository.store(
+    byID: storeID,
+    eventLoop: eventLoop,
+    logger: logger,
+    baggage: baggage
+)
+```
+
+### Explicitly passing `Logger` always
+
+Imagine we don't have metadata providers, we'd have to manually set trace IDs on loggers which doesn't really work as all libraries involved would need to know about the same specific trace ID. Event if we inverted the dependency to have `Tracing` depend on `Logging` so that we'd be able to define something like "Tracing, please populate this logger with metadata", we'd have to make sure this thing is called in all places to avoid dropping contextual metadata.
```

Comment:
```suggestion
Imagine we don't have metadata providers, we'd have to manually set trace IDs on loggers which doesn't really work as all libraries involved would need to know about the same specific trace ID. Even if we inverted the dependency to have `Tracing` depend on `Logging` so that we'd be able to define something like "Tracing, please populate this logger with metadata", we'd have to make sure this thing is called in all places to avoid dropping contextual metadata.
```

Reply:
Not everything needs to be event-driven 😜